\section{Contribution}
In this chapter, we discuss the design and implementation of
practical mechanisms for
collaborative two-party computation on genome data.  Our main focus is
on the dynamic programming algorithms such as the edit distance and
the Smith-Waterman algorithm for sequence alignment, which are among
the fundamental building blocks of computational biology~\cite[Chapter
11]{Gusfield}.

Specifically, the following contributions are made:

\begin{itemize}

\item 
We design and implement several efficient, privacy-preserving protocols
for computing the \emph{edit distance} between two strings $\alpha$
and $\beta$, \ie, the minimum number of {\sf delete}, {\sf insert},
and {\sf replace} operations needed to convert $\alpha$ into $\beta$.

\item
We construct an efficient solution for computing the Smith-Waterman~\cite{Smith-Waterman}
similarity score between two sequences. The Smith-Waterman similarity score is used for
sequence alignment.

\item
We demonstrate that, in addition to privacy-preserving computation on
genomic data, our techniques generalize to a wide variety of dynamic
programming problems~\cite[Chapter 15]{CLR}.

\item
We evaluate our implementation on realistic case studies, including
protein sequences from the Pfam database~\cite{pfam2002}.  Our experimental results
demonstrate that our methods are practical on sequences of up to several
hundred symbols in length.

\item
Even though theoretical constructions for various secure multi-party
computation (SMC) tasks have received much attention (see related
work below), actual implementations and performance measurements
are exceptionally rare.  In contrast to the vast majority of
SMC research, the implementations of the protocols presented here have been
evaluated on real protein sequences and analysis workloads, demonstrating
that they can be used in practice on problem instances of realistic size,
while achieving the same level of cryptographic security as theoretical
constructions.  We also make all of the implementation available as open source
software.\footnote{\url{http://www.cs.wisc.edu/~lpkruger/thesis}}

\end{itemize}

\section{Threats}
Genomic data such as DNA and protein sequences are increasingly collected
by government agencies for law enforcement and medical purposes,
disseminated via public repositories for research and medical studies,
and even stored in the private databases of commercial enterprises.
For example, deCODE Genetics aims to collect the complete genome sequences
of the entire population of Iceland~\cite{decode}, while the non-profit
HapMap Project is developing a public repository of representative genome
sequences in order to help researchers to discover genes associated with
specific diseases~\cite{hapmap}.

The underlying genome records are typically collected from specific
individuals, and thus contain a lot of sensitive personal information,
including genetic markers for diseases, information that can be be used to
establish paternity and maternity, and so on.  Therefore, genomic records
are usually stored in an anonymized form, that is, without explicit
references to the identities of people from whom they were collected.

Even if genome sequences are anonymized, \emph{re-identification} is a
major threat.  In many cases, a malicious user can easily de-anonymize
the sequence and link it to its human contributor simply by recognizing
the presence of certain markers~\cite{cmugenome}.  Furthermore,
many genetic markers are expressible in the person's phenotype, which
includes externally observable features~\cite{harvard}.  In general,
protecting privacy of individual DNA when the corresponding genome
sequence is available to potential attackers does not appear realistic.
Developing practical tools which would support collaborative analysis of
genomic data without requiring the participants to release the underlying
DNA and protein sequences is perhaps the most important privacy challenge
in computational biology today.

\section{Related work:} 
%Public availability of personal information due to the Internet has
%brought privacy concerns to the forefront~\cite{cra99,tur03}.  Therefore,
%there has been considerable interest in developing privacy protection
%technologies~\cite{p3p02,gwb97,rks+97}.

One of the fundamental cryptographic primitives for designing
privacy-preserving protocols is \emph{secure function evaluation (SFE)},
introduced in chapter 2.  Generic constructions, however, are not always
practical in real-world
application scenarios.  For example, direct application of SFE to the
computational biology problems is infeasible on anything but toy examples.

Special-purpose privacy-preserving protocols have been developed
for tasks such as auctions, surveys, remote diagnostics, and so
on~\cite{FPRS04,FNP04,LP02,NPS99,BPS07}, but privacy-preserving
genomic computation has received little attention.  We are aware of
only two papers devoted to this or similar problems: Atallah \textit{et
al.}~\cite{atallah} and Szaida \textit{et al.}~\cite{Szajda:NDSS:2004}.

Neither paper provides a proof of security.  The edit distance protocol
of~\cite{atallah} is impractical even for very small problem instances due
to its immense computational cost: a single iteration requires at least
300,000 modular multiplications (see section~\ref{appendix-atallah}).
The distributed Smith-Waterman algorithm of~\cite{Szajda:NDSS:2004}
involves decomposing the problem instance into several sub-problems,
which are spread out to several participants.  The intuition is that
each participant solves a sub-problem and thus it is presumed that
he cannot infer the inputs for the original problem (this does not
imply proper cryptographic security).  It is unclear how the protocol
of~\cite{Szajda:NDSS:2004} may be used in the two-party case, or whether
it can be generalized to other dynamic programming algorithms.

By contrast, our techniques are provably secure and substantially more
scalable, as demonstrated by our evaluation on realistic instances of
genomic analysis problems.

For testing equality between two values, our constructions use
Yao's garbled circuits method presented in 
section~\ref{sub:Garbled-Circuit-Method}, albeit in
a non-``black-box'' way (our edit distance protocol exploits the
specifics of circuit encoding in Yao's method).  Related work on
equality testing includes Damg{\aa}rd \textit{et al.}~\cite{Dam-TCC06}.
Their techniques guarantee unconditional security, but only allow a
minority of passive (semi-honest) adversaries, and thus do not apply in
the two-party setting of our protocol.  We will use as a building
block, the Pinkas-Naor oblivious transfer protocol presented
in section~\ref{sub:Oblivious-Transfer}.
